### PR TITLE
Add support for relaying a batch of calls to L2 via the governance bridge

### PR DESCRIPTION
**Pull request by i-stam**

### CODE REVIEW

**Comment by liamzebedee** [comment]
Hey @i-stam, dumb q, can the L1-L2 messenger relay `msg.value`?

File: contracts/OwnerRelayOnEthereum.sol [hunk 1]
```diff
@@ -62,7 +66,31 @@ contract OwnerRelayOnEthereum is MixinSystemSettings, Owned {
         emit RelayInitiated(target, data);
     }
 
+    function initiateRelayBatch(
+        address[] calldata targets,
+        bytes[] calldata data,
```
Comment (ajsantander) [inline contracts/OwnerRelayOnEthereum.sol:71]: Maybe call these `datas`?

File: contracts/OwnerRelayOnOptimism.sol [hunk 2]
```diff
@@ -71,13 +66,25 @@ contract OwnerRelayOnOptimism is MixinResolver, TemporarilyOwned {
...
+        iAbs_BaseCrossDomainMessenger messenger = _messenger();
+
+        require(msg.sender == address(messenger), "Sender is not the messenger");
+        require(messenger.xDomainMessageSender() == _ownerRelayOnEthereum(), "L1 sender is not the owner relay");
+
+        for (uint256 i = 0; i < targets.length; i++) {
+            _relayCall(targets[i], data[i]);
+        }
+
+        emit CallBatchRelayed(targets, data);
```
Comment (ajsantander) [inline contracts/OwnerRelayOnOptimism.sol:83]: I think I'd prefer having 3 events like so:
* `RelayFinalized`
* `RelayBatchFinalized`
* `DirectRelay`

Would make it more consistent with `OwnerRelayOnEthereum`.

File: contracts/OwnerRelayOnEthereum.sol [hunk 3]
```diff
@@ -62,7 +66,31 @@ contract OwnerRelayOnEthereum is MixinSystemSettings, Owned {
...
+    ) external onlyOwner {
+        // First check that the length of the arguments match
+        require(targets.length == data.length, "Argument length mismatch");
+
+        IOwnerRelayOnOptimism ownerRelayOnOptimism;
+        bytes memory messageData = abi.encodeWithSelector(ownerRelayOnOptimism.finalizeRelayBatch.selector, targets, data);
+
+        // Use specified crossDomainGasLimit if specified value is not zero.
+        // otherwise use the default in SystemSettings.
+        uint32 xGasLimit =
```
Comment (ajsantander) [inline contracts/OwnerRelayOnEthereum.sol:82]: Can we avoid duplicating this code and perhaps abstract it to a private utility function?

File: contracts/OwnerRelayOnOptimism.sol [hunk 4]
```diff
@@ -71,13 +66,25 @@ contract OwnerRelayOnOptimism is MixinResolver, TemporarilyOwned {
...
+
+        emit CallRelayed(target, data);
     }
 
-    function directRelay(address target, bytes calldata data) external onlyTemporaryOwner {
-        _relayCall(target, data);
+    function finalizeRelayBatch(address[] calldata targets, bytes[] calldata data) external {
+        iAbs_BaseCrossDomainMessenger messenger = _messenger();
+
+        require(msg.sender == address(messenger), "Sender is not the messenger");
```
Comment (ajsantander) [inline contracts/OwnerRelayOnOptimism.sol:76]: I think we can also dedup these two requires.

File: contracts/OwnerRelayOnEthereum.sol (hunk 1 above)
Comment (ajsantander) [reply contracts/OwnerRelayOnEthereum.sol:71]: Or `payloads`, but make sure to make it consistent everywhere else.

**Comment by i-stam** [comment]
> Hey @i-stam, dumb q, can the L1-L2 messenger relay `msg.value`?

@liamzebedee ,  good question. I guess the only way to do it is via an implicit param i.e. get `msg.value` and wrap it as param on L2.

File: contracts/OwnerRelayOnEthereum.sol (hunk 1 above)
Comment (i-stam) [reply contracts/OwnerRelayOnEthereum.sol:71]: fixed

File: contracts/OwnerRelayOnEthereum.sol (hunk 3 above)
Comment (i-stam) [reply contracts/OwnerRelayOnEthereum.sol:82]: fixed

File: contracts/OwnerRelayOnOptimism.sol (hunk 4 above)
Comment (i-stam) [reply contracts/OwnerRelayOnOptimism.sol:76]: fixed

File: contracts/OwnerRelayOnOptimism.sol (hunk 2 above)
Comment (i-stam) [reply contracts/OwnerRelayOnOptimism.sol:83]: addressed

File: contracts/OwnerRelayOnEthereum.sol [hunk 5]
```diff
@@ -9,6 +10,8 @@ import "@eth-optimism/contracts/iOVM/bridge/messaging/iAbs_BaseCrossDomainMessen
 
 interface IOwnerRelayOnOptimism {
     function finalizeRelay(address target, bytes calldata data) external;
+
+    function finalizeRelayBatch(address[] calldata target, bytes[] calldata data) external;
```
Comment (ajsantander) [inline contracts/OwnerRelayOnEthereum.sol:14]: Replace `data` with `payload`
and `data` with `payloads`

File: contracts/OwnerRelayOnOptimism.sol [hunk 6]
```diff
@@ -35,13 +32,23 @@ contract OwnerRelayOnOptimism is MixinResolver, TemporarilyOwned {
...
-    function _relayCall(address target, bytes memory data) private {
+    function _relayCall(address target, bytes memory payload) private {
         // solhint-disable avoid-low-level-calls
-        (bool success, bytes memory result) = target.call(data);
+        (bool success, bytes memory result) = target.call(payload);
 
         require(success, string(abi.encode("xChain call failed:", result)));
+    }
+
+    function onlyAllowMessengerAndL1Relayer() internal view {
```
Comment (ajsantander) [inline contracts/OwnerRelayOnOptimism.sol:42]: Name this function with an underscore for consistency?

**Review by ajsantander — APPROVED** [review]
What a beauty! Good work @i-stam 🙌

File: contracts/OwnerRelayOnOptimism.sol (hunk 6 above)
Comment (i-stam) [reply contracts/OwnerRelayOnOptimism.sol:42]: right, I missed this, good catch

**Pull request merged**